lib/AtomicCache.php

```diff
@@ -0,0 +1,176 @@
+<?php
+
+namespace Amp\Cache;
+
+use Amp\Promise;
+use Amp\Sync\KeyedMutex;
+use Amp\Sync\Lock;
+use function Amp\call;
+
+final class AtomicCache implements Cache
+{
+    /** @var Cache */
+    private $cache;
+
+    /** @var KeyedMutex */
+    private $mutex;
+
+    public function __construct(Cache $cache, KeyedMutex $mutex)
+    {
+        $this->cache = $cache;
+        $this->mutex = $mutex;
+    }
+
+    /**
+     * Attempts to get the value for the given key. If the value is not found, the key is locked, the $create callback
+     * is invoked with the key as the first parameter. The value returned from the callback is stored in the cache and
+     * the promise returned from this method is resolved with the value.
+     *
+     * @param string   $key
+     * @param callable(string $key) $create
+     * @param int|null $ttl
+     *
+     * @return Promise<mixed>
+     *
+     * @throws CacheException If the $create callback throws an exception while generating the value.
+     */
+    public function getOrSet(string $key, callable $create, ?int $ttl = null): Promise
```

Comment:
Maybe name this computeIfAbsent?

Reply:
getOrSet is a common name for such a method.

Reply:
getOrCreate may be better?

Reply:
Symfony uses simply `get`. In Nette framework which I used before it's called `load`.

Reply:
I guess I like `load` as well.